### PR TITLE
Feature 1965 ioda2nc same input

### LINKED ISSUE #1965: Fix IODA2NC to handle the same input file being provided multiple times
*Replace italics below with details for this issue.*

## Describe the Problem ##
ioda2nc fails if the same input file is given with -iodafile option

### Expected Behavior ###
It should work without errors.

### Environment ###
Describe your runtime environment:
*1. Machine: (Linux Workstation)*
*2. OS: (RedHat Linux)*
*3. Software version number(s): 10.1.0 and 11.0 beta*

### To Reproduce ###
Describe the steps to reproduce the behavior:
*1. Go to seneca*
*2. run the following command*

/usr/local/met/bin/ioda2nc -v 2 /d1/projects/METplus/METplus_Data/development/feature_1203_ioda2nc/met_test/new/ioda \
  /ioda.NC001007.2020031012.nc ioda.NC001007.2020031012.summary.nc \
 -iodafile /d1/projects/METplus/METplus_Data/development/feature_1203_ioda2nc/met_test/new/ioda/ioda.NC001007.2020031012.nc

*3. See error*

terminate called after throwing an instance of 'netCDF::exceptions::NcEdge'
  what():  NetCDF: Start+count exceeds dimension bound
file: ncVar.cpp  line:958
Aborted

*4. run the following command*

export METPLUS_ELEVATION_RANGE_DICT="elevation_range = {beg = -1000;end = 100000;}"
export METPLUS_LEVEL_RANGE_DICT=""
export METPLUS_MASK_DICT=""
export METPLUS_MESSAGE_TYPE=""
export METPLUS_MESSAGE_TYPE_GROUP_MAP=""
export METPLUS_MESSAGE_TYPE_MAP=""
export METPLUS_METADATA_MAP=""
export METPLUS_MET_CONFIG_OVERRIDES=""
export METPLUS_MISSING_THRESH=""
export METPLUS_OBS_NAME_MAP="obs_name_map = [{ key = \"wind_direction\"; val = \"WDIR\"; }, { key = \"wind_speed\";     val = \"WIND\"; }];"
export METPLUS_OBS_VAR=""
export METPLUS_OBS_WINDOW_DICT="obs_window = {beg = -5400;end = 5400;}"
export METPLUS_QUALITY_MARK_THRESH="quality_mark_thresh = 0;"
export METPLUS_STATION_ID=""
export METPLUS_TIME_SUMMARY_DICT="time_summary = {flag = TRUE;raw_data = TRUE;beg = \"000000\";end = \"235959\";step = 300;width = 600;grib_code = [];obs_var = [\"WIND\"];type = [\"min\", \"max\", \"range\", \"mean\", \"stdev\", \"median\", \"p80\"];vld_freq = 0;vld_thresh = 0.0;}"
export MET_TMP_DIR="/d1/personal/mccabe/out2/tmp"

/usr/local/met/bin/ioda2nc -v 2 /d1/projects/METplus/METplus_Data/development/feature_1203_ioda2nc/met_test/new/ioda \
  /ioda.NC001007.2020031012.nc ioda.NC001007.2020031012.summary.nc \
  -config /d1/personal/mccabe/METplus/parm/met_config/IODA2NCConfig_wrapped \
 -iodafile /d1/projects/METplus/METplus_Data/development/feature_1203_ioda2nc/met_test/new/ioda/ioda.NC001007.2020031012.nc

*5. See error*

ERROR  :
ERROR  : get_obs_data_float() -> WDIR@ObsValue does not exist!
ERROR  :

*Post relevant sample data following these instructions:*
*https://dtcenter.org/community-code/model-evaluation-tools-met/met-help-desk#ftp*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2702691

### Assignee ###
- [ ] Select **engineer(s)** or **no engineer** required
- [x] Select **scientist(s)** or **no scientist** required: no scientist

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [ ] Select **Organization** level **Project** for support of the current coordinated release
- [ ] Select **Repository** level **Project** for development toward the next official release or add **alert: NEED PROJECT ASSIGNMENT** label
- [ ] Select **Milestone** as the next bugfix version

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)

## Bugfix Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [ ] Fork this repository or create a branch of **main_\<Version>**.
Branch name: `bugfix_<Issue Number>_main_<Version>_<Description>`
- [ ] Fix the bug and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **main_\<Version>**.
Pull request: `bugfix <Issue Number> main_<Version> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Organization** level software support **Project** for the current coordinated release
Select: **Milestone** as the next bugfix version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Complete the steps above to fix the bug on the **develop** branch.
Branch name:  `bugfix_<Issue Number>_develop_<Description>`
Pull request: `bugfix <Issue Number> develop <Description>`
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by hsoh-u**
- The duplicated variable names are filtered.
- The output variable names from the config file are processed just before saving NetCDF file
- Use the raw variable names for internal processing.

## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

Ran two commands: with/without -iodafile 

Command 1: without -iodafile  option (no error)

/d1/personal/hsoh/git/pull_request/MET_feature_1965_ioda2nc_same_input/met/bin/ioda2nc /d1/personal/hsoh/data/IODA_files/ioda.NC001007.2020031012.nc out_ioda.NC001007.2020031012.nc_units.nc /d1/projects/MET/MET_regression/develop/NB20210916/MET-develop/test/config/IODA2NCConfig_mask -v 2

Command 2: with -iodafile and the same input file name  (failed to generate output before)

/d1/personal/hsoh/git/pull_request/MET_feature_1965_ioda2nc_same_input/met/bin/ioda2nc /d1/personal/hsoh/data/IODA_files/ioda.NC001007.2020031012.nc out_ioda.NC001007.2020031012.nc_units.nc /d1/projects/MET/MET_regression/develop/NB20210916/MET-develop/test/config/IODA2NCConfig_mask -iodafile /d1/personal/hsoh/data/IODA_files/ioda.NC001007.2020031012.nc -v 2

The problem was the same variable names were added twice into the variable name list.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Do the similar test with different input file or run unit test for ioda2nc.
Or run METplus script you have.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

Added one more unit test, "ioda2nc_same_input".

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

One more output file from the unit test: ioda2nc/ioda.NC001007.2020031012.same_input.nc

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the PR definition above.
- [ ] Ensure the PR title matches the feature or bugfix branch name.
- [ ] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [ ] After submitting the PR, select **Linked issues** with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: met/src/tools/other/ioda2nc/ioda2nc.cc [hunk 1]
```diff
@@ -775,7 +776,14 @@ void process_ioda_file(int i_pb) {
...
       for(idx=0; idx<v_obs_data.size(); idx++ ) {
-         obs_arr[1] = idx;
+         int var_idx;
+         if (!obs_var_names.has(raw_var_names[idx], var_idx)) {
+            mlog << Warning << "\n" << method_name
+                 << "Skip the variable " << raw_var_names[idx]
+                 << " at " << ioda_files[i_pb] << "\n\n";
+            continue;
+         }
+         obs_arr[1] = var_idx;
```
Comment (georgemccabe) [inline met/src/tools/other/ioda2nc/ioda2nc.cc:786]: Is var_idx uninitialized at this point?

**Review by georgemccabe — CHANGES_REQUESTED** [review]
I ran some tests using the feature branch executable on kiowa and confirmed that the expected behavior occurred. I modified the IODA2NC.conf METplus use case conf file to include a duplicate input file. I ran the use case using this branch and confirmed that it ran successfully and skipped the duplicate file. I ran the use case again on the latest MET beta on kiowa and confirmed that it fails.

I made one inline comment about a variable that appears to be uninitialized when it is used unless it is being set by the command in the if-block that is run before the line in question. Once that is fixed or if I am misunderstanding what is happening here, then I approve of this pull request.

File: met/src/tools/other/ioda2nc/ioda2nc.cc (hunk 1 above)
Comment (hsoh-u) [reply met/src/tools/other/ioda2nc/ioda2nc.cc:786]: It's set by the next line: "if (!obs_var_names.has(raw_var_names[idx], var_idx))" but it 's safe to initialized before updating it. It's corrected.

Comment (georgemccabe) [reply met/src/tools/other/ioda2nc/ioda2nc.cc:786]: I looked into the StringArray::has function and it looks like this logic sets var_idx to the index where the value was found. Sorry, I misunderstood how the function worked.

**Review by georgemccabe — APPROVED** [review]
I misunderstood the logic in my first review. The changes look good.

**Pull request merged**